### PR TITLE
Factor out CRAMReferenceSource as an interface and fix two URI bugs.

### PR DESCRIPTION
**Pull request by cmnbroad**
### Description

Fixes https://github.com/samtools/htsjdk/issues/597. Factors out CRAMReferenceSource as an interface so that alternate implementations can be provided by consumers such as IGV. (Now that we have this as an interface, there is additional factoring that can be done i.e., separate the ENA CRAM service ReferenceSource implementation from the the file-based ReferenceSource implementation but we'll save those for a separate PR).

Also fixes https://github.com/samtools/htsjdk/issues/331 and https://github.com/samtools/htsjdk/issues/332.

For  https://github.com/samtools/htsjdk/issues/331, this PR adds a method overload that accepts a URI (though this is really just syntactic sugar). This also fixes the specific issue described in https://github.com/samtools/htsjdk/issues/332, though TribbleIndexedFeatureReader may have other issues using parameterized URLs (details in that ticket - trying to keep this minimal for now in the interest of getting IGV unblocked).

### Checklist

Note: the changes to the CRAM ReferenceSource are not backward compatible, though for most consumers they should be mostly source compatible).

- [x] Code compiles correctly
- [x] New tests covering changes and new functionality
- [x] All tests passing
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

**Comment by droazen** [comment]
@jrobinso Could you have a look at this, and let us know if everything you requested is here?

**Comment by jrobinso** [comment]
This looks complete for now.   When its merged I will do a pull and start the IGV part.

File: src/java/htsjdk/samtools/CRAMIterator.java [hunk 1]
```diff
@@ -94,10 +93,10 @@ public CRAMIterator(final InputStream inputStream, final ReferenceSource referen
         parser = new ContainerParser(cramHeader.getSamFileHeader());
     }
 
-    public CRAMIterator(final SeekableStream seekableStream, final ReferenceSource referenceSource, final long[] coordinates, final ValidationStringency validationStringency)
+    public CRAMIterator(final SeekableStream seekableStream, final CRAMReferenceSource referenceSource, final long[] coordinates, final ValidationStringency validationStringency)
             throws IOException {
         if (null == referenceSource) {
-            throw new CRAMException("A reference source is required for CRAM files");
+        //    throw new CRAMException("A reference source is required for CRAM files");
```
Comment (droazen) [inline src/java/htsjdk/samtools/CRAMIterator.java]: Why did you comment out this line? It seems to me that we still want to throw if `referenceSource` is null.

File: src/java/htsjdk/samtools/SamReaderFactory.java [hunk 2]
```diff
@@ -78,7 +79,7 @@
     abstract public SamReaderFactory referenceSequence(File referenceSequence);
 
     /** Sets the specified reference sequence * */
-    abstract public SamReaderFactory referenceSource(ReferenceSource referenceSequence);
+    abstract public SamReaderFactory referenceSource(CRAMReferenceSource referenceSequence);
```
Comment (droazen) [inline src/java/htsjdk/samtools/SamReaderFactory.java]: It seems awkward that this abstract class `SamReaderFactory` should have CRAM-specific parameters/return values like this -- though I suppose that was always the case, just disguised under the former, more generic name.

File: src/java/htsjdk/samtools/cram/ref/CRAMReferenceSource.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+    /**
+     * getReferenceBases
+     * @param sequenceRecord the SAMSequenceRecord identifying the reference
+     *                       being requested
+     * @param tryNameVariants if true, attempt to match the requested sequence name
+     *                        against the reference by using common name variations,
+     *                        such as adding or removing a leading "chr" prefix
+     *                        from the requested name. if false, use exact match
+     * @return the bases representing the requested sequence
```
Comment (droazen) [inline src/java/htsjdk/samtools/cram/ref/CRAMReferenceSource.java]: Document method behavior when the reference sequence can't be found.

File: src/java/htsjdk/samtools/cram/ref/ReferenceSource.java [hunk 4]
```diff
@@ -40,14 +40,14 @@
 import java.util.Map;
 import java.util.regex.Pattern;
 
-public class ReferenceSource {
+public class ReferenceSource implements CRAMReferenceSource {
```
Comment (droazen) [inline src/java/htsjdk/samtools/cram/ref/ReferenceSource.java]: Argh...in a just world `CRAMReferenceSource` would implement `ReferenceSource`, not vice versa. Can we rename `ReferenceSource` to something like `CRAMReferenceSourceImpl` to avoid confusion?

File: src/java/htsjdk/tribble/TribbleIndexedFeatureReader.java [hunk 5]
```diff
@@ -271,6 +273,27 @@ private void readHeader() throws IOException {
         return new WFIterator();
     }
 
+    //Visible for testing
+    static boolean isGZIPPath(final String path) {
```
Comment (droazen) [inline src/java/htsjdk/tribble/TribbleIndexedFeatureReader.java]: Should this method be moved to a utils class, instead of being buried inside here?

File: src/tests/java/htsjdk/tribble/TribbleIndexFeatureReaderTest.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        return new Object[][]{
+                {TestUtils.DATA_DIR + "test.vcf", 5},
+                {TestUtils.DATA_DIR + "test.vcf.gz", 5}
+        };
+    }
+
+    @Test(dataProvider = "featureFileStrings")
+    public void testIndexedGZIPVCF(final String testPath, final int expectedCount) throws IOException {
+        final VCFCodec codec = new VCFCodec();
+        final TribbleIndexedFeatureReader<VariantContext, LineIterator> featureReaderHttp =
```
Comment (droazen) [inline src/tests/java/htsjdk/tribble/TribbleIndexFeatureReaderTest.java]: How is this an "http" feature reader? It's reading from local files, isn't it?

File: src/tests/java/htsjdk/tribble/TribbleIndexFeatureReaderTest.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                new TribbleIndexedFeatureReader(testPath, codec, false);
```
Comment (droazen) [inline src/tests/java/htsjdk/tribble/TribbleIndexFeatureReaderTest.java]: Close reader via try-with-resources

File: src/tests/java/htsjdk/tribble/TribbleIndexFeatureReaderTest.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        };
+    }
+
+    @Test(enabled = true, dataProvider = "blockCompressedExtensionExtensionURIStrings")
+    public void testGZExtension(final String testString, final boolean expected) throws URISyntaxException {
+        Assert.assertEquals(TribbleIndexedFeatureReader.isGZIPPath(testString), expected);
+    }
+
+    @DataProvider(name = "featureFileStrings")
+    public Object[][] createGZExtensionStringss() {
```
Comment (droazen) [inline src/tests/java/htsjdk/tribble/TribbleIndexFeatureReaderTest.java]: `Stringss` -> `Strings`

**Comment by droazen** [comment]
Review complete, back to @cmnbroad, merge after addressing comments.

File: src/java/htsjdk/samtools/CRAMIterator.java (hunk 1 above)
Comment (cmnbroad) [inline src/java/htsjdk/samtools/CRAMIterator.java]: Hmm. Done.

File: src/java/htsjdk/samtools/cram/ref/CRAMReferenceSource.java (hunk 3 above)
Comment (cmnbroad) [inline src/java/htsjdk/samtools/cram/ref/CRAMReferenceSource.java]: Documented to return null, since that is what the current sole implementation does.

File: src/java/htsjdk/samtools/cram/ref/ReferenceSource.java (hunk 4 above)
Comment (cmnbroad) [inline src/java/htsjdk/samtools/cram/ref/ReferenceSource.java]: Agreed. we'll do this in a separate PR.

File: src/java/htsjdk/tribble/TribbleIndexedFeatureReader.java (hunk 5 above)
Comment (cmnbroad) [inline src/java/htsjdk/tribble/TribbleIndexedFeatureReader.java]: There are similar functions in IOUtil, but they accept another extension for gzip, and its unclear to me what to what extent these should overlap or not. So I'm leaving this here as a narrow extension to the local test that was previously being done.
